memCache/entryCache.go

```diff
@@ -433,3 +215,216 @@ func (e *Entry) timeToUpdate() bool {
 	}
 	return false
 }
+
+type EntryCache struct {
+	log            *log.Logger
+	clk            clock.Clock
+	requestTimeout time.Duration
+	entries        map[string]*Entry   // one-to-one map keyed on name -> entry
+	lookupMap      map[[32]byte]*Entry // many-to-one map keyed on sha256 hashed OCSP requests -> entry
+	StableBackings []stableCache.Cache
+	issuers        *issuerCache
+	client         *http.Client
+	mu             sync.RWMutex
+}
+
+func NewEntryCache(clk clock.Clock, logger *log.Logger, monitorTick time.Duration, stableBackings []stableCache.Cache, client *http.Client, timeout time.Duration, issuers []*x509.Certificate) *EntryCache {
+	c := &EntryCache{
+		log:            logger,
+		entries:        make(map[string]*Entry),
+		lookupMap:      make(map[[32]byte]*Entry),
+		StableBackings: stableBackings,
+		client:         client,
+		requestTimeout: timeout,
+		clk:            clk,
+		issuers:        newIssuerCache(issuers),
+	}
+	go c.monitor(monitorTick)
+	return c
+}
+
+func hashEntry(h hash.Hash, name, pkiBytes []byte, serial *big.Int) ([32]byte, error) {
+	issuerNameHash, issuerKeyHash, err := common.HashNameAndPKI(h, name, pkiBytes)
+	if err != nil {
+		return [32]byte{}, err
+	}
+	serialHash := sha256.Sum256(serial.Bytes())
+	return sha256.Sum256(append(append(issuerNameHash, issuerKeyHash...), serialHash[:]...)), nil
+}
+
+func allHashes(e *Entry) ([][32]byte, error) {
```

Comment:
It looks to me like these hashes are used for indexing into lookupMap, but you have full control over what you look things up by. Why would you need multiple hashes here?